spec/requests/billboards_spec.rb

```diff
@@ -7,18 +7,89 @@
   let(:max_age) { 15.minutes.to_i }
   let(:stale_if_error) { 26_400 }
 
+  def create_billboard(**options)
+    defaults = {
+      approved: true,
+      published: true,
+      display_to: :all
+    }
+    create(:display_ad, **options.reverse_merge(defaults))
+  end
+
   describe "GET /:username/:slug/billboards/:placement_area" do
-    before do
-      create(:display_ad, placement_area: "post_comments", published: true, approved: true)
-    end
+    let!(:billboard) { create_billboard(placement_area: "post_comments") }
 
     it "returns the correct response" do
       get article_billboard_path(username: article.username, slug: article.slug, placement_area: "post_comments")
-      display_ad = DisplayAd.find_by(placement_area: "post_comments")
 
       expect(response).to have_http_status(:ok)
-      expect(response.parsed_body).to include(display_ad.processed_html)
+      expect(response.parsed_body).to include(billboard.processed_html)
+    end
+
+    # rubocop:disable RSpec/NestedGroups
+    context "when client geolocation is present" do
+      let(:client_in_alberta_canada) { { "HTTP_CLIENT_GEO" => "CA-AB" } }
+      let(:client_in_california_usa) { { "HTTP_CLIENT_GEO" => "US-CA" } }
+
+      before do
+        allow(FeatureFlag).to receive(:enabled?).with(:billboard_location_targeting).and_return(true)
+      end
+
+      context "with signed-in user" do
+        let!(:canada_billboard) { create_billboard(placement_area: "sidebar_left", target_geolocations: "CA") }
+        let!(:us_billboard) { create_billboard(placement_area: "sidebar_left", target_geolocations: "US") }
+
+        before do
+          sign_in user
+        end
+
+        it "returns only billboards targeting their location" do
+          # DisplayAd.for_display uses random sampling, so we collect a few for confidence
+          5.times do
+            get article_billboard_path(username: article.username, slug: article.slug, placement_area: "sidebar_left"),
+                headers: client_in_alberta_canada
+
+            expect(response).to have_http_status(:ok)
+            expect(response.parsed_body).to include(canada_billboard.processed_html)
+            expect(response.parsed_body).not_to include(us_billboard.processed_html)
+
+            get article_billboard_path(username: article.username, slug: article.slug, placement_area: "sidebar_left"),
```

Comment:
I would split the tests for `client_in_alberta_canada` and `client_in_california_usa`: 2 `get`s and 3 expects is worth a separate test.

I'm also a little reluctant to run specs 5 times though I get the random sampling argument, but still there are only 2 billboards available (just a thought, not require to change anything)

Reply:
Maybe we could use the `priority` attribute here to help surface the appropriate billboard instead of running the specs 5 times?

Reply:
I thought about it and since there are only two billboards the test would be ridiculously flaky if the implementation really didn't work (failing about half the time) so it's probably okay to just have a normal spec? I have updated them